src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHeader.Write.cs

```diff
@@ -609,18 +649,38 @@ private int WriteGnuFields(Span<byte> buffer)
         private static void WriteData(Stream archiveStream, Stream dataStream, long actualLength)
         {
             dataStream.CopyTo(archiveStream); // The data gets copied from the current position
+            WriteEmptyPadding(archiveStream, actualLength);
+        }
 
+        // Calculates the padding for the current entry and writes it after the data.
+        private static void WriteEmptyPadding(Stream archiveStream, long actualLength)
+        {
             int paddingAfterData = TarHelpers.CalculatePadding(actualLength);
             if (paddingAfterData != 0)
             {
                 Debug.Assert(paddingAfterData <= TarHelpers.RecordSize);
 
-                Span<byte> padding = stackalloc byte[TarHelpers.RecordSize];
-                padding = padding.Slice(0, paddingAfterData);
-                padding.Clear();
+                Span<byte> zeros = stackalloc byte[TarHelpers.RecordSize];
+                zeros = zeros.Slice(0, paddingAfterData);
+                zeros.Clear();
+
+                archiveStream.Write(zeros);
+            }
+        }
+
+        // Calculates the padding for the current entry and asynchronously writes it after the data.
+        private static ValueTask WriteEmptyPaddingAsync(Stream archiveStream, long actualLength, CancellationToken cancellationToken)
+        {
+            int paddingAfterData = TarHelpers.CalculatePadding(actualLength);
+            if (paddingAfterData != 0)
+            {
+                Debug.Assert(paddingAfterData <= TarHelpers.RecordSize);
 
-                archiveStream.Write(padding);
+                byte[] zeros = new byte[paddingAfterData];
+                return archiveStream.WriteAsync(padding, cancellationToken);
```

Comment:
nit: this needs a rename to fix the compiler error

```suggestion
                return archiveStream.WriteAsync(zeros, cancellationToken);
```